cpatch: Search base image when patching files.

jagombar opened the pull request:
Often the differences between a base container image and the build being used as the cpatch source result in a new image that fails to bootstrap a cluster.  This is usually caused by incompatibilities between the binaries and libraries in the new image or (in the case of centos 9 based images) by python packages being patched into the wrong directory for the base image)

This PR changes cpatch to ensure that a more complete set of binaries and libraries are replaced in the base image.
It initially changed only the shell script version of cpatch and then when the python replacement was discovered, the same changes were made to that script.

Both versions of the script were tested with multiple components selected to patch and with no components selected (equivalent to --all)

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

jagombar commented on src/script/cpatch:
```diff
@@ -195,11 +193,13 @@ if [ $rgw -eq 1 ] || [ $all -eq 1 ]; then
     dockerfile+=$'ADD rgw /usr/bin\n'
 
-    RGWLIBS="libradosgw.so*"
+    RGWLIBS="librados.so.* libceph-common.so.*"
```
I meant to add a comment about this change.  libradosgw.so doesn't exist any more and the script fails if you specify --all.  It was removed almost 2 years ago as far as I could tell.  Looking at the ldd output for radosgw, these were the two ceph libraries that were required.

jagombar commented on src/script/cpatch:
```diff
@@ -195,11 +193,13 @@ if [ $rgw -eq 1 ] || [ $all -eq 1 ]; then
-    for f in lib/$RGWLIBS; do cp $f $TMP/rgwlib ; done
+    for f in $RGWLIBS; do cp lib/$f $TMP/rgwlib ; done
     [ $strip -eq 1 ] && strip $TMP/rgwlib/*
     dockerfile+=$'ADD rgwlib /usr/lib64\n'
```
I realised that this was duplicating work if the --all parameter was specified, but struggled to find a neat way to avoid it.  I like the idea of copying all binaries and libraries to the same place and then moving them at the end so I'll make that change,

bill-scales reviewed (commented):
Latest changes resolve all my comments

phlogistonjohn commented on src/script/cpatch.py:
```diff
@@ -511,79 +528,37 @@ def _copy_binary(self, src_path, dst_path):
-            log.warning("skipping lib %s: %s", name, err)
-        return f"ADD {name} {destination}"
-
-    def _py_site_packages(self):
```
I'm utterly confused by why you're removing steps, esp. ones related to python.

jagombar replied:
latest commit restores _py_site_packages()

phlogistonjohn reviewed (commented):
It's cool to see someone else using cpatch.py, but I'm severely struggling to understand the changes.

Could you try breaking the cpatch.py related changes into multiple parts that can be reviewed piece by peice? Thanks.

jagombar commented:
I've split the changes to the python version into 3 separate commits.  Hopefully this makes the changes easier to understand and review.

I don't think that I've removed any of the steps relating to python.  The _py_site_packages() function is no longer needed because the Dockerfile now contains steps to search for the location of the files to be patched (using "find").

phlogistonjohn commented on src/script/cpatch.py:
```diff
@@ -613,7 +580,9 @@ def _py_common_job(self, component):
                 )
-        return [f"ADD {name} {self._py_site_packages()}"]
+        dlines=[f"ADD {name} tmp_python_common"]
+        dlines.append ("RUN for i in tmp_python_common/*; do find /usr/lib/python*/site-packages -type d -name $(basename $i) -exec cp -frpv $i/* '{}' \;; done && rm -rf tmp_python_common")
```
Sorry, but I don't care for this approach. It feels like embedding a full-blown shell script inside a dockerfile inside a python script. On top of that, I intentionally limited the list of python versions to find to list of actual python versions.

jagombar replied:
We used this approach to find the correct location of the binaries and libraries because there is a longer list of sub directories where the files to patch might be located (depending on the distribution used in the base container).  
The alternative approach of running the "ls -d" command inside a container inside a python script would not easily work for these files because there is a longer list of sub directories to search and multiple directories might exist with the ceph binaries only in one of them.  
It seems preferable to use the same approach for finding all files to patch (python, binaries and libraries).  In addition, limiting the directories searched to python versions that exist, means that the script needs continual updates.

jagombar replied:
reverted to original behaviour for python files

phlogistonjohn commented on src/script/cpatch.py:
```diff
@@ -513,39 +529,29 @@ def _copy_binary(self, src_path, dst_path):
         out = []
 
         bin_src = self._ctx.build_dir / "bin"
-        bin_dst = self._workdir / f"{prefix}_bin"
```
I don't understand why this needs to change. Why isn't prefix just set to "tmp" where appropriate?

jagombar replied:
I changed this because prefix was always being set to "tmp" so it didn't seem to add any value.
All binaries end up in the tmp_bin directory because that is where the Dockerfile build process will search for them when it patches the existing files in the container.  
I can revert this if you would prefer.

phlogistonjohn replied:
> I can revert this if you would prefer.

Yes please. Let's make the changes as small and targeted as possible.

jagombar replied:
done

phlogistonjohn reviewed (commented):
Please note that I will not be reviewing the shell script changes. I had originally wanted cpatch.py to replate the cpatch shell script but we had trouble determining if there was anyone using the script or not and decided not to risk breaking any one existing work flows.  But I rewrote it in python so I would not have to look at the shell version again ;-)

I have a better sense of what's going on here now, but still puzzled by a number of things.  However, I will admit that I don't use cpatch.py for C/C++ components - only python stuff - myself. I don't really think the python stuff needs changing at this time. Can we focus on only fixing the C/C++ related things for now? Can you explain what's going wrong with the C/C++ parts so I have a better idea of what the problem being solved is?

jagombar commented:
The shell script version of cpatch is used by the cstart script.  Both of these are referenced in [https://github.com/ceph/ceph/blob/main/doc/dev/cephadm/developing-cephadm.rst](url) which is why we initially changed the shell script version, before discovering the python replacement.

The patching of binaries and libraries was broken because the files could be located in multiple locations depending on the distribution and version of the base container.  For example the script was putting libceph-common in /usr/lib64 instead of /usr/lib64/ceph/.  The former directory was maybe correct for some earlier container images but the latter directory was the correct location for the library with centos 9 based container images.  
Having the script search for the current directory of each file in the image and then replace the file at that location is a more flexible and future proof approach.

jagombar commented:
OK, updated, rebased and retested.

jagombar commented:
@phlogistonjohn are you happy with the python file changes now?  

I'm not sure what to do with the commit that changes the shell script version.  I could remove that commit from this PR and leave it broken which seems a shame or maybe try and find somebody else to review the changes to that file.

phlogistonjohn commented:
Sorry, but I still see `_py_site_packages` getting removed. For now, I'd strongly prefer not to see any changes to the python stuff that AFICT work, as I use it almost every day. I'll be more flexible on the C/C++ related stuff because I don't hack on that stuff myself.

As for the cpatch shell script, I'd try pinging someone from core, or ask around on Slack.

perezjosibm commented on src/script/cpatch.py:
```diff
@@ -639,60 +645,39 @@ def _core_job(self, component):
             bin_patterns=["radosgw", "radosgw-admin"],
             lib_patterns=["librados.so*", "libceph-common.so*"],
         )
         return out
```
This looks like a copy/paste mistake, the second return is not needed (unreachable code as indicated by the linter).

jagombar replied:
Thanks Jose, well spotted, but this isn't a problem introduced by my changes and I don't want to start making additional changes - just trying to keep this as small as possible now.

jagombar commented:
Sorry for the delay.  I've removed the commit that changes the python site_packages stuff, but need to retest the script again so have put this back to draft for now.

jagombar commented:
@phlogistonjohn I've reverted the behaviour for the patching of the python files in cpatch.py and retested the script.  Please could you take another look when you have a moment.

phlogistonjohn reviewed (approved):
Works for me, thanks for sticking with it and being receptive to my requests. Approved for cpatch.py changes.

jdurgin reviewed (approved):
Nice improvements! This approach looks more future-proof to changes in the container / packaging.
